docs/specs/user-model.md

```diff
@@ -1,71 +1,87 @@
-          * [Validation of Email Login ID](#validation-of-email-login-id)
-          * [Normalization of Email Login ID](#normalization-of-email-login-id)
-          * [Unique key generation of Email Login ID](#unique-key-generation-of-email-login-id)
-        * [Username Login ID](#username-login-id)
-          * [Validation of Username Login ID](#validation-of-username-login-id)
-          * [Normalization of Username Login ID](#normalization-of-username-login-id)
-          * [Unique key generation of Username Login ID](#unique-key-generation-of-username-login-id)
-        * [Phone Login ID](#phone-login-id)
-          * [Validation of Phone Login ID](#validation-of-phone-login-id)
-          * [Normalization of Phone Login ID](#normalization-of-phone-login-id)
-          * [Unique key generation of Phone Login ID](#unique-key-generation-of-phone-login-id)
-        * [Raw Login ID](#raw-login-id)
-      * [Optional Login ID Key during authentication](#optional-login-id-key-during-authentication)
-      * [The purpose of unique key](#the-purpose-of-unique-key)
-  * [Authenticator](#authenticator)
-    * [Primary Authenticator](#primary-authenticator)
-    * [Secondary Authenticator](#secondary-authenticator)
-    * [Authenticator Tags](#authenticator-tags)
-    * [Authenticator Types](#authenticator-types)
-      * [Password Authenticator](#password-authenticator)
-      * [TOTP Authenticator](#totp-authenticator)
-      * [OOB-OTP Authenticator](#oob-otp-authenticator)
-    * [Device Token](#device-token)
-    * [Recovery Code](#recovery-code)
-
 ## User
 
-A user has many identities. A user has many authenticators.
+A user has a set of custom attributes.
+The custom attributes can contribute to the computation of the claims of the user.
+
+A user has many identities.
+Identity has [Standard Claims](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims).
+The claims of an identity can contribute to the computation of the claims of the user.
+
+A user has many authenticators.
+The claims of an authenticators do **NOT** contribute to the computation of the claims of the user.
+
+A user has a set of claims, which
+
+- Are computed from the custom attributes and the claims of the identities.
+- Are the information returned in the UserInfo endpoint.
+- Are included in the ID Token.
+
+## Custom Attributes
+
+Custom Attributes is a JSON Object. The developer can update it via Admin API.
+
+Custom Attributes can be any valid JSON Object. There is no limitation on the nesting level. However, a sensible size limit (e.g. 10MiB) is enforced.
+
+The developer can save whatever they want in Custom Attributes.
+
+> NOTE: The initial implementation is very likely to be a single textarea for editing the whole custom attributes object.
+> Much more fancier features like generated form fields requires us to interpret the JSON schema and generate form fields.
+> That could be a very useful standalone product.
```

Comment:
The value can be complex values like JSON object and JSON array, which does not fit well in a text field.

Reply:
@louischan-oursky Any better suggestions that's not very complicated to develop in our MVP, but also not a textarea that is impossible for normal users?

Or we can have key textfields, and each value is textarea

Reply:
@louischan-oursky can we resolve this problem in the specification before finalizing it?

Reply:
The custom attributes have a JSON schema defined the developer.
An example of similar thing is AWS Policy editor. AWS policy has schema as well. They have a wizard to generate policy as well as a plain old textarea for editing the policy JSON directly.

Since the provided JSON schema could be very complex and it is non-trivial to take a schema and generate form fields, I prefer we just stick with a textarea in our MVP. It ensures we provide the full capability of custom attributes editing experience.

On the other hand, how likely would a "normal user" edit the custom attributes? Custom attributes initially should be written by the developer via Admin API. The developer can store something tightly related to their business logic, which may require more application knowledge to make changes.

Reply:
> On the other hand, how likely would a "normal user" edit the custom attributes? 

My guess is very often, because:

1. Developer may use the custom attributes as some sort of customer profile, where business users will need to look at it or update it from management portal.
2. Following 1, it might even some time handy for developer to just enable changing it from our setting pages (this one might be a bit unclear how good / bad it is)

Hence I think without it might vastly limited the usage of custom attributes...

However I'm okay if we implement it in multiple phase, like phase 1 MVP only text area, followed by phase 2 make it view and changeable by business user in management portal; After that we can stop and wait for a few months with more real users, and see if they even want it in setting page.